lib/promise.rb

```diff
@@ -107,24 +155,74 @@ def wait
     end
   end
 
+  def state
+    @state || :pending
+  end
+
   protected
 
   # Override to defer calling the callback for Promises/A+ spec compliance
   def defer
     yield
   end
 
-  def add_callback(callback)
-    if pending?
-      @callbacks << callback
-      callback.source = self
+  def settle_from_handler(value)
+    fulfill(yield(value))
+  rescue => ex
+    reject(ex)
+  end
+
+  def target
+    @target ? @target.target : self
+  end
+
+  def add_callback(callback, on_fulfill_arg, on_reject_arg)
+    @callbacks ||= []
+    @callbacks.push(callback, on_fulfill_arg, on_reject_arg)
+  end
+
+  def migrate_callbacks(callbacks)
+    @callbacks.concat(callbacks)
+  end
+
+  def promise_fulfilled(value, on_fulfill)
+    if on_fulfill
+      settle_from_handler(value, &on_fulfill)
     else
-      dispatch!(callback)
+      fulfill(value)
+    end
+  end
+
+  def promise_rejected(reason, on_reject)
+    if on_reject
+      settle_from_handler(reason, &on_reject)
+    else
+      reject(reason)
     end
   end
 
   private
 
+  def fulfill_promises
+    return unless @callbacks
+
+    @callbacks.each_slice(3) do |callback, on_fulfill_arg, _|
+      defer { callback.send(:promise_fulfilled, @value, on_fulfill_arg) }
```

Comment:
I don't think the semantics around this are clearly specified. I tried moving the `defer` call into `#promise_fulfilled` / `#promise_rejected`, but that broke some of the existing test cases.

A promise that was resolved by a different promise always used the source promise's `defer` call. Maybe we could add a check to ensure that we only migrate callbacks across promise classes that are actually the same?